### ISSUE TITLE
Globals, table & memory metadata not exposed to JS

### ISSUE THREAD

**Issue by brion** [WebAssembly/binaryen#3573]
I'm working on an experimental async wasm interpreter, piggybacking off binaryen.js to parse the binary and let me walk the AST. For the most part this is working well so far, but there's a few things that are missing from the JS API:

* Globals can be retrieved only via a reference from an export, or via a get/set expression. There's no way to walk the list of all globals, which I would need to initialize values without walking the entire codebase first.
* Memory segment info can be read, but not info about the memory itself such as its initial and maximum sizes, whether it's shared, or whether it's imported.
* Table imports are visible, but the initial and maximum sizes are also not exposed.

I'll try to whip up a provisional patch adding these (until then I can work around it by making certain assumptions that won't hold with all modules).

**Comment by kripken** [comment]
A patch would be great!

Note that in the long term I hope to automate our JS API bindings. All the necessary info should be in declarative form here:

https://github.com/WebAssembly/binaryen/blob/main/src/wasm-delegations-fields.h

A patch to add specific things would be fine, but if you add more than one or two things it might actually be faster to write the autogenerator.

**Comment by brion** [comment]
I'll take a peek, thanks!

**Comment by brion** [comment]
@kripken
> Note that in the long term I hope to automate our JS API bindings. All the necessary info should be in declarative form here:
> 
> https://github.com/WebAssembly/binaryen/blob/main/src/wasm-delegations-fields.h
> 
> A patch to add specific things would be fine, but if you add more than one or two things it might actually be faster to write the autogenerator.

It looks like those delegation macros would allow replacing the implementation of `Module.getExpressionInfo` without having to manually add every new expression id. However I think this is separate from other metadata like looking at the attached globals, tables, and memories. I think the implementation of one could proceed separately from the implementation of the other and they would not interfere with each other.

**Comment by kripken** [comment]
@brion Oh, yes, that's right. The proposed autogeneration stuff would be for expressions, but module level things are separate, at least for now.

**Comment by brion** [comment]
Cool, well I might or might not get to that too but that'll be for another day. :D

**Comment by brion** [comment]
@kripken while I'm thinking on it though one more thing -- the TypeScript annotations in the binaryen.js node package are nice as they help with code completion in Visual Studio Code; it'd be nice to figure a way to start generating them as well. These need to be static information, not dynamic code, so the delegate callbacks won't work unless code is run on a list of all possible expressions, which perhaps could be rigged up but feels like an extra step.

Another possibility is to generate the function calls in wasm-delegations-fields.h from a static data structure, which can also be used to generate the TypeScript annotations.

**Comment by brion** [comment]
Ok, looks like there's a couple more things hiding in the current code that let me get at globals and table info which I'll need to push a patch to the TypeScript definitions over on the other repo for, but in the meantime I'm almost done with some additional memory info funcs that's now enough to instantiate a WebAssembly.Memory with correct initial and maximum sizes.

I'll push a PR tomorrow or so after testing it some more. :)

**Comment by kripken** [comment]
@brion Yes, TS bindings would be good to do as well. And yes, it would need an extra step I think - a small tool written to use the delegates header, that generates C++ that generates the TS, basically.

**Comment by brion** [comment]
@kripken cool, I'll look at the TS bindings later as well as the other bits. In the meantime, PR specifically addressing the missing access to memory info is now passing tests so I don't seem to have broken anything at least. ;)

https://github.com/WebAssembly/binaryen/pull/3595

It's working well so far in my interpreter, which needs to know the data to instantiate a live memory to attach to the interpreted function wrappers.

The necessary functions for enumerating globals and tables were already in there, though the separate TS bindings are out of date and don't include them which I'll clean up later. :)
